src/Microsoft.OpenApi.Readers/V2/OpenApiDocumentDeserializer.cs

```diff
@@ -119,21 +119,48 @@ internal static partial class OpenApiV2Deserializer
             {s => s.StartsWith("x-"), (o, p, n) => o.AddExtension(p, n.CreateAny())}
         };
 
-        private static void MakeServers(IList<OpenApiServer> servers, ParsingContext context)
+        private static void MakeServers(IList<OpenApiServer> servers, ParsingContext context, Uri defaultUrl)
         {
             var host = context.GetFromTempStorage<string>("host");
             var basePath = context.GetFromTempStorage<string>("basePath");
             var schemes = context.GetFromTempStorage<List<string>>("schemes");
 
+            // If nothing is provided, don't create a server
+            if (host == null && basePath == null && schemes == null)
+            {
+                return;
+            }
+
+            // Fill in missing information based on the defaultUrl
+            host = host ?? defaultUrl.GetComponents(UriComponents.NormalizedHost, UriFormat.SafeUnescaped);
```

Comment:
>defaultUrl.GetComponents(UriComponents.NormalizedHost, UriFormat.SafeUnescaped); [](start = 27, length = 80)

V2 spec:

If the host is not included, the host serving the documentation is to be used (including the port). 

This is in line with V3 spec which defaults server URL to "/" with the condition that "This URL supports Server Variables and MAY be relative, to indicate that the host location is relative to the location where the OpenAPI document is being served."


We have a few special cases to handle:

=> Host empty + schemes not empty =>  This is weird. I don't know how to represent this in V3. Maybe we can simply ignore the schemes since we don't really have a way to handle it.

=> Host empty + schemes empty + basePath not empty => Essentially, what this means is we use the Host serving the doc but with appended path. To represent this in V3, we simply need the basePath in Server.url

=> Host empty + schemes empty + basePath empty => You handled this. This means no need for Server object in V3.

=> Host not empty + schemes empty => We can just use the host and basePath without schemes in Server.url in V3? That still seems like a valid URL to me.

Reply:
Just so we are the same page, the BaseUrl in the Settings is where you put "the Host serving the doc".  I think what you are saying, is you don't want an arbitrary default for "host serving the doc" when one isn't provided.  Is that correct?

Reply:
ah okay. Correct - it's a little odd to default to an arbitrary example.org. Also, the conversion is possible (except case 1&2) above even without the info on "host serving the doc", so I'm not sure if we really need it.

---
In reply to: [193228718](https://github.com/Microsoft/OpenAPI.NET/pull/267#discussion_r193228718) [](ancestors = 193228718)

Reply:
Ah. I see that the baseUrl has to be there for reference resolving. In that case, I think we should leave it blank like you suggested, handle the MakeServers here, and use the property for Reference Resolving as planned.

---
In reply to: [193231179](https://github.com/Microsoft/OpenAPI.NET/pull/267#discussion_r193231179) [](ancestors = 193231179,193228718)